File: frontEnd/intermediateAlgorithm/romanNumConverter.js
/*jslint node: true */
'use strict';

function convertToRoman (num) {
  var result = '';
  var ones, tens, hunds, thous;
  var hundsA = ['C', 'CC', 'CCC', 'CD', 'D', 'DC', 'DCC', 'DCCC', 'CM'];
  var tensA = ['X', 'XX', 'XXX', 'XL', 'L', 'LX', 'LXX', 'LXXX', 'XC'];
  var onesA = ['I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX'];
  thous = Math.floor(num / 1000);
  hunds = Math.floor(num % 1000 / 100);
  tens = Math.floor(num % 100 / 10);
  ones = Math.floor(num % 10);
  console.log(thous, hunds, tens, ones);
  for (var i = 0; i < thous;i++) {
    result += 'M';
  }
  if (hunds) {
    result += hundsA[hunds - 1];
  }
  if (tens) {
    result += tensA[tens - 1];
  }
  if (ones) {
    result += onesA[ones - 1];
  }
  return result;
}

console.log(convertToRoman(36));
console.log(convertToRoman(68));
console.log(convertToRoman(97));
console.log(convertToRoman(1000));
console.log(convertToRoman(3999));

File: 2-algorithm/intermediateGo/romanNumConverter.go
package main

import (
	"fmt"
	"math"
)

func convertToRoman(n int) string {
	var res string
	hundreds := []string{"C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"}
	tens := []string{"X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"}
	units := []string{"I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"}
	t := int(math.Floor(float64(n / 1000)))
	h := int(math.Floor(float64(n % 1000 / 100)))
	d := int(math.Floor(float64(n % 100 / 10)))
	u := int(math.Floor(float64(n % 10)))
	for i := 0; i < t; i++ {
		res += "M"
	}
	if h > 0 {
		res += hundreds[h-1]
	}
	if d > 0 {
		res += tens[d-1]
	}
	if u > 0 {
		res += units[u-1]
	}
	return res
}

func romanNumConverter() {
	fmt.Println(convertToRoman(36))
	fmt.Println(convertToRoman(68))
	fmt.Println(convertToRoman(97))
	fmt.Println(convertToRoman(1000))
	fmt.Println(convertToRoman(3999))
}
